### PR TITLE
rfc: Cloud Assembly Specification

### PR DESCRIPTION
**Pull request by RomainMuller**
Looking forward to supporting a more integrated deployment experience
with support of multiple stacks (with deployment order constraints
honored) as well as assets, this proposes a specification for a
*Cloud Assembly* that determines a document storage format.

Fixes: #956

---

TODO:
* [ ] Specify the deployment behavior (order resolution, ability to parallelize, ...)
* [ ] Improve the phrasing around *missing information*
* [X] Add TypeScript definition for the manifest document (must be possible to generate JSON Schema from it)
* [ ] Add language around Drop providers publishing a JSON schema that allows for easy validation
* [ ] Describe how deployment systems should proceed
* [ ] Flesh out the "full assembly" example
* [X] Describe digital signature algorithm
* [ ] Refine the digital signature validation procedure & requirements

### CODE REVIEW

File: specifications/cloud_assembly.md [hunk 1]
```diff
@@ -0,0 +1,390 @@
+# Cloud Assembly Specification, Version 1.0
+
+The key words **MUST**, **MUST NOT**, **REQUIRED**, **SHALL**, **SHALL NOT**, **SHOULD**, **SHOULD NOT**,
+**RECOMMENDED**, **MAY**, and **OPTIONAL** in this document are to be interpreted as described in [RFC 2119] when they
+are spelled out in bold, capital letters (as they are shown here).
+
+## Introduction
+A *Cloud Assembly* is a document container fail designed to hold the components of *cloud-native* applications,
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: "self-contained document container"... 🤔

File: specifications/cloud_assembly.md [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+including all the parts that are needed in order to deploy those to *the cloud*. This document exposes the specification
+of the *Cloud Assembly* format as well as requirements imposed on *Cloud Assemblers* and *Cloud Runtimes*.
+
+### Goals
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Well I very much would like the specification to be fundamentally cloud-agnostic, which is why I use "the cloud" and not "AWS". As for extensibility, this may not be the right term indeed (settled for this now until I figure out a better one).

File: specifications/cloud_assembly.md [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+* A *Cloud Assembly* is self-contained, making deployments reproductible.
+
+## Specification
+A *Cloud Assembly* is a ZIP archive that **SHOULD** conform to the [ISO/IEC 21320-1:2015] *Document Container File*
+standard. Use of the `deflate` compression method is **RECOMMENDED** in order to minimize the size of the resulting
+file. *Cloud Assembly* files **SHOULD** use the `.cloud` extension in order to make them easier to recognize by users.
+
+Documents in the archive can be stored with any name and directory structure, however the following entries at the root
+of the archive are reserved for special use:
+* `manifest.json` **MUST** be present and contains the [manifest document](#manifest-document) for the *Cloud Assembly*.
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: "If it's a cat, call it a cat".

File: specifications/cloud_assembly.md [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+Key           |Type                 |Required|Description
+--------------|---------------------|:------:|-----------
+`schema`      |`string`             |Required|The schema for the document. **MUST** be `cloud-assembly/1.0`.
+`drops`       |`Map<ID,Drop>`       |Required|A mapping of [*Logical ID*](#logical-id) to [`Drop`](#drop).
+`missing`     |`Map<string,Missing>`|        |A mapping of context keys to [missing information](#missing).
+
+The [JSON] specification allows for keys to be specified multiple times in a given `object`. However, *Cloud Assembly*
+consumers **MAY** assume keys are unique, and [*Cloud Assemblers*](#cloud-assemblers) **SHOULD** avoid generating
+duplicate keys. If duplicate keys are present, the latest specified value **SHOULD** be preferred.
+
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Agreed, that was always the plan :P

File: specifications/cloud_assembly.md [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+  + Upper-case letters: `A` (`0x41`) through `Z` (`0x5A`)
+  + Lower-case letters: `a` (`0x61`) through `z` (`0x7A`)
+  + Numeric characters: `0` (`0x30`) through `9` (`0x39`)
+  + Plus: `+` (`0x2B`)
+  + Minus: `-` (`0x2D`)
+  + Forward-slash: `/` (`0x2F`)
+  + Underscore: `_` (`0x5F`)
+
+### `Drop`
+`Drop`s are the building blocks of *Cloud Assemblies*. They model a part of the *cloud-native* application that can be
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: I like a consistent markup for words with a particular meaning. But if you think that makes too much clutter...

[the remark this answers is not in the document]
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Where "quite a lot" is "twice" :) But I'm all for reducing required acumen.

File: specifications/cloud_assembly.md [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+Key          |Type             |Required|Description
+-------------|-----------------|:------:|-----------
+`type`       |`string`         |Required|The [*Drop Type*](#drop-type) specifier of this `Drop`.
+`environment`|`string`         |required|The [environment](#environment) specifier for this `Drop`.
+`metadata`   |`Map<string,any>`|        |Arbitrary key-value pairs associated with this `Drop`.
+`properties` |`Map<string,any>`|        |The properties of this `Drop` as documented by its maintainers.
+
+Each [`Drop` Type](#drop-type) can produce outputs that can be used in order to allow other `Drop`s to consume the
+resources they procude. Each `Drop` implementer is responsible to document the output attributes it supports. References
+to these outputs are modeled using special `string` tokens within entries of the `properties` section of `Drop`s:
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: That's a part of it but I reckon there might be more to it than just this. It'll be pretty inefficient to discuss over GitHub issues, though. Let's talk about this & report the outcome back here.

File: specifications/cloud_assembly.md [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+
+#### Drop Type
+Every `Drop` has a type specifier, which allows *Cloud Assembly* consumers to know how to deploy it. The type specifiers
+are `string`s that use an URI-like syntax (`protocol://path`), providing the coordinates to a reference implementation
+for the `Drop` behavior.
+
+Deployment systems **MUST** support at least one protocol, and **SHOULD** support all the protocols specified in
+the following sub-sections.
+
+##### The `npm` protocol
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: That's what I was thinking but I wanted to talk about it before writing it.

File: specifications/cloud_assembly.md [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+The design goals for the *Cloud Assembly Specification* are the following:
+* The *Cloud Assembly Specification* is extensible.
+* The *Cloud Assembly Specification* is cloud-agnostic.
+* The *Cloud Assembly Specification* is easy to implement and use.
+* The *Cloud Assembly Specification* supports authenticity and integrity guarantees.
+* A *Cloud Assembly* is self-contained, making deployments reproductible.
+
+## Specification
+A *Cloud Assembly* is a ZIP archive that **SHOULD** conform to the [ISO/IEC 21320-1:2015] *Document Container File*
+standard. Use of the `deflate` compression method is **RECOMMENDED** in order to minimize the size of the resulting
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: This seems unnecessarily specific and also out of scope for this spec. For example, the JAR specification doesn't seem to have suffered from not telling people that they should pick a particular compression method.

File: specifications/cloud_assembly.md [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+Key            |Type             |Required|Description
+---------------|-----------------|:------:|-----------
+`provider`     |`string`         |Required|A tag that identifies the entity that should provide the information.
+`props`        |`Map<string,any>`|Required|Properties that are required in order to obtain the missing information.
+
+### Digital Signature
+#### Signing
+*Cloud Assemblers* **SHOULD** support digital signature of *Cloud Assemblies*. When support for digital signature is
+present, *Cloud Assemblers*:
+* **MUST** require the user to specify which [PGP][RFC 4880] key should be used.
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: Who is "the user" here?

Comment (rix0rrr) [reply specifications/cloud_assembly.md]: Again, this section feels somewhat premature to me.

File: specifications/cloud_assembly.md [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+The `manifest.json` file is the entry point of the *Cloud Assembly*. It **MUST** be a valid [JSON] document composed of
+a single `object` that conforms to the following schema:
+
+Key           |Type                 |Required|Description
+--------------|---------------------|:------:|-----------
+`schema`      |`string`             |Required|The schema for the document. **MUST** be `cloud-assembly/1.0`.
+`drops`       |`Map<ID,Drop>`       |Required|A mapping of [*Logical ID*](#logical-id) to [Drop](#drop).
+`missing`     |`Map<string,Missing>`|        |A mapping of context keys to [missing information](#missing).
+
+The [JSON] specification allows for keys to be specified multiple times in a given `object`. However, *Cloud Assembly*
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: Sooo... is this a long way to say "do whatever JSON does" ?

File: specifications/cloud_assembly.md [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+* `signature.asc`, when present, **MUST** contain the [digital signature](#digital-signature) of the *Cloud Assembly*.
+
+### Manifest Document
+The `manifest.json` file is the entry point of the *Cloud Assembly*. It **MUST** be a valid [JSON] document composed of
+a single `object` that conforms to the following schema:
+
+Key           |Type                 |Required|Description
+--------------|---------------------|:------:|-----------
+`schema`      |`string`             |Required|The schema for the document. **MUST** be `cloud-assembly/1.0`.
+`drops`       |`Map<ID,Drop>`       |Required|A mapping of [*Logical ID*](#logical-id) to [Drop](#drop).
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: The name "drop" was not at all obvious to me. My first impression was that it would be a set of logical IDs that would be removed ("dropped") from templates somehow. Why not a more obvious name like "component", "artifact", "deployable"...?

File: specifications/cloud_assembly.md [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+  cross-drop references to separate the *Logical ID* from the *attribute* name.
+
+In other words, *Logical IDs* are expected to match the following regular expression:
+```js
+/^[A-Za-z0-9+\/_-]{1,256}$/
+```
+
+### Drop
+Clouds are made of Drops. Thet are the building blocks of *Cloud Assemblies*. They model a part of the
+*cloud application* that can be deployed independently, provided it's dependencies are fulfilled. Drops are specified
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: typo its

File: specifications/cloud_assembly.md [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+using [JSON] objects that **MUST** conform to the following schema:
+
+Key          |Type             |Required|Description
+-------------|-----------------|:------:|-----------
+`type`       |`string`         |Required|The [*Drop Type*](#drop-type) specifier of this Drop.
+`environment`|`string`         |required|The target [environment](#environment) in which Drop is deployed.
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: "is" deployed? "should be" deployed?

File: specifications/cloud_assembly.md [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+`metadata`   |`Map<string,any>`|        |Arbitrary key-value pairs associated with this Drop.
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: Would be worthwhile defining some standard metadata.

For example, **warnings** and **errors** seem very useful to standardize, as well as the behavior associated with them (errors would block deployment, warnings would cause a prompt, etc.)

File: specifications/cloud_assembly.md [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+  ╰───┬───╯ ╰─────┬─────╯
+      │           └─ The name of the output attribute
+      └───────────── The Logical ID of the Drop
+```
+
+The following escape sequences are valid:
+* `\\` encodes the `\` literal
+* `\${` encodes the `${` literal
+
+Deployment systems **SHOULD** return an error upon encountering an occurrence of the `/` literal that is not part of a
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: `\`?

File: specifications/cloud_assembly.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (rix0rrr) [reply specifications/cloud_assembly.md]: Both of these are tying declaration to implementation.

More generically, how about just an identifier string? (Could be namespaced a la `com.amazonaws.cloudformationstack` or something) 

In that case, we can even build tools to statically analyze the drops, without necessarily executing them. In direct opposition is the following:

```
{
   "type": "bash -c '$(which foobar)/../libexec/bin/cfnprovider'"
}
```

How is any tool going to analyze all the CloudFormation stacks in an Assembly if this is the format of "identifier" they may have to contend with?

A compromise might be found in the following:

```
{
  "type": "com.amazonaws.cloudformation.stack",
  "executionHint": "cdk-cloudformation-stack"
}
```

File: specifications/cloud_assembly.md [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+Property    |Type    |Required|Description
+------------|--------|:------:|-----------
+`file`      |`string`|Required|The assembly-relative path to the file that will be uploaded.
+`bucketName`|`string`|Required|The name of the bucket where this file will be uploaded.
+`objectKey` |`string`|Required|The key at which to place the object in the bucket.
+
+##### Output Attributes
+Attribute   |Type    |Description
+------------|--------|-----------
+`bucketName`|`string`|The name of the bucket where the file was uploaded.
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: So, taking the output here would purely be to force sequencing of operations, right?

File: specifications/cloud_assembly.md [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+------------|--------|:------:|-----------
+`savedImage`|`string`|Required|The assembly-relative path to the tar archive obtained from `docker image save`.
+`imageName` |`string`|Required|The name of the image (e.g: `000000000000.dkr.ecr.bermuda-triangle-1.amazon.com/name`).
+`tagName`   |`string`|        |The name of the tag to use when pushing the image (default: `latest`).
+
+##### Output Attributes
+Attribute     |Type    |Description
+--------------|--------|-----------
+`exactImageId`|`string`|An absolute reference to the published image version (`imageName@DIGEST`).
+`imageName`   |`string`|The full tagged image name (`imageName:tagName`).
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: Make this name different from input `imageName` so that it's clear it's not the same schema. `fullImageName`? `qualifiedImageName`?

File: specifications/cloud_assembly.md [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
+`objectKey` |`string`|The key at which the file was uploaded in the bucket.
+
+##### Example
+```json
+{
+  "type": "npm://@aws-cdk/assets.FileDrop",
+  "environment": "aws://000000000000/bermuda-triangle-1",
+  "properties": {
+    "file": "assets/file.bin",
+    "bucket": "${helperStack.bucketName}",
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: What if the `helperStack` is the bootstrap stack? Does a CDK app always add in the bootstrap stack into the assembly, or is the bootstrap stack supplied as a parameter to the assembly?

File: specifications/cloud_assembly.md [hunk 19]
```diff
@@ -0,0 +1,398 @@
...
+
+## Annex
+### Examples of Drops for the AWS Cloud
+#### `@aws-cdk/aws-cloudformation.StackDrop`
+A [*CloudFormation* stack][CFN Stack].
+
+##### Properties
+Property     |Type                |Required|Description
+-------------|--------------------|:------:|-----------
+`template`   |`string`            |Required|The assembly-relative path to the *CloudFormation* template document.
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: `stackName` needs to be a parameter as well.

Or will you force `stackName` == logicalID of `drop`?

File: specifications/cloud_assembly.md [hunk 20]
```diff
@@ -0,0 +1,398 @@
...
+`parameters` |`Map<string,string>`|        |Parameters to be passed to the [stack][CFN Stack] upon deployment.
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: Is the only way to order drops by using the `output` of one in the `parameters` of another? We have more need of sequencing stack drops that aren't necessarily expressed as a threading of information from one object to the other (yay, side effects).

For example, an `Export` in one stack and an `Import` in the other.

I would argue for a field similar to CloudFormation's `DependsOn`.

File: specifications/cloud_assembly.md [hunk 21]
```diff
@@ -0,0 +1,398 @@
...
+          "websiteFilesKeyPrefix": "${StaticFiles.keyPrefix}",
+        }
+      }
+    },
+    "DockerImage": {
+      "type": "npm://@aws-cdk/aws-ecr.DockerImageDrop",
+      "environment": "aws://123456789012/eu-west-1",
+      "properties": {
+        "savedImage": "docker/docker-image.tar",
+        "imageName": "${PipelineStack.ecrImageName}"
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: According to your schema of values returned by `StackDrop` this ought to be `${PipelineStack.outputs.ecrImageName}`.

File: specifications/cloud_assembly.md [hunk 22]
```diff
@@ -0,0 +1,398 @@
...
+      }
+    },
+    "StaticFiles": {
+      "type": "npm://@aws-cdk/assets.DirectoryDrop",
+      "environment": "aws://123456789012/eu-west-1",
+      "properties": {
+        "directory": "assets/static-website",
+        "bucketName": "${PipelineStack.stagingBucket}"
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: Same here.

File: specifications/cloud_assembly.md [hunk 23]
```diff
@@ -0,0 +1,398 @@
...
+}
+```
+
+#### `@aws-cdk/aws-ecr.DockerImageDrop`
+A Docker image to be published to an *ECR* registry.
+
+##### Properties
+Property    |Type    |Required|Description
+------------|--------|:------:|-----------
+`savedImage`|`string`|Required|The assembly-relative path to the tar archive obtained from `docker image save`.
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: Can we also have a `localImageName` for a simplified workflow of `docker build + docker push`, instead of a `docker build + docker save + docker push`?

I'd like to save copying 1GB of data around on every build if it's not necessary.

File: specifications/cloud_assembly.md [hunk 24]
```diff
@@ -0,0 +1,398 @@
...
+`imageName` |`string`|Required|The name of the image (e.g: `000000000000.dkr.ecr.bermuda-triangle-1.amazon.com/name`).
```
Comment (rix0rrr) [inline specifications/cloud_assembly.md]: I'd prefer for this name to be something like `pushTarget` or something.

File: specifications/cloud_assembly.md (hunk 8 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: OTOH `.class` files don't necessarily compress particularly well. But yeah, maybe this is going a little overboard.

File: specifications/cloud_assembly.md (hunk 9 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: I wanna specify signature early - it's dear to my heart.

File: specifications/cloud_assembly.md (hunk 10 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: It says "JSON allows whatever, but you SHOULD NOT use that trait of it" - insisting that re-using the same key multiple times results in undefined behavior.

File: specifications/cloud_assembly.md (hunk 11 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Would you have been less fooled if `Drop` had been enclosed in backticks, making it look more like a technical name?

File: specifications/cloud_assembly.md (hunk 13 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: "should" sounds like a recommendation, where it's a requirement. I can switch to "will be" ?

File: specifications/cloud_assembly.md (hunk 14 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: I agree. Although IMO, "errors" should prevent the cloud assembly from even being created...

File: specifications/cloud_assembly.md (hunk 15 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Yas, of course.

File: specifications/cloud_assembly.md (hunk 16 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Correct. There is also an option where we may not get a bucketName input, and instead the provisioning tool figures it out for us (so the parameter might become optional). I'm not entirely sure yet.

File: specifications/cloud_assembly.md (hunk 17 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Would `imageId` make sense here? I guess it collides unfortunately with docker terminology...

File: specifications/cloud_assembly.md (hunk 18 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: So this is the case where you explicitly specify "everything" in your cloud assembly, but in case your "provisioning orchestrator" does this for you, then it wouldn't be part of the assembly. That's where the optional inputs I referred to in a previous reply come into play - the provisioning orchestrator wires it's own values in at deploy time, and they are available for reading from the asset `Drop`'s outputs.

File: specifications/cloud_assembly.md (hunk 19 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Good point. I think I was implicitly assuming logicalId = stackName, but that feels brittle, especially if you want to computer-generate the logical ID's (the spec, after all, doesn't bind a particular meaning to them beyond "they identify a thing with the doc").

File: specifications/cloud_assembly.md (hunk 20 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Absolutely!

File: specifications/cloud_assembly.md (hunk 21 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Oh yeah. That might need to change. Good catch!

File: specifications/cloud_assembly.md (hunk 24 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Yeah these serve mostly as example, so names in here are definitely non-normative and I've been... favoring speed over quality :D

File: specifications/cloud_assembly.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (rix0rrr) [reply specifications/cloud_assembly.md]: 1) I was responding to Elad's suggestion of using subprocesses here.

2) Even if we were to use `npm://` coordinates, does the identifier have meaning without the NPM package that contains code to interpret it? To have any hope of tools that interpret the model apart from simply executing it, it must.

File: specifications/cloud_assembly.md (hunk 11 above)
Comment (rix0rrr) [reply specifications/cloud_assembly.md]: No. It's that Elad referred to a pun in a different comment, but I had completely missed the punnability of the word "drop".

I think the better term in that case is "droplet", which I think is more commonly understood as a noun that refers to a tiny sphere of water.

Or a more obvious cloud pun like "Cumulus" or whatever?

File: specifications/cloud_assembly.md (hunk 14 above)
Comment (rix0rrr) [reply specifications/cloud_assembly.md]: A single Cloud Assembly can contain multiple drops, one of which contains synthesis errors. That need not prohibit deployment of the other drops.

Or something similar. But we need a way to express this, see https://github.com/awslabs/aws-cdk/issues/1035

File: specifications/cloud_assembly.md (hunk 23 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Added a short blurb to note that the AWS Drop specifications provided in the document are for illustration purpose only. The API design for those can be a separate process :) But I hear the concern and it's valid, I just don't want to overload the Cloud Assembly spec with those details (at least not right now).

File: specifications/cloud_assembly.md (hunk 20 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: The Export/Import is a good case. Although one could say you might want to use Parameters to make the dependency explicit (but you certainly don't have to). I'm adding an optional `dependsOn` array.

File: specifications/cloud_assembly.md (hunk 14 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: To me this feels akin to your linker produces a binary that includes some static libraries that had completion errors, and you're asking for the executable to be produced, so you can... maybe attempt to use the other features? I would argue that drops that contain errors should then... be omitted from the assembly entirely.

File: specifications/cloud_assembly.md (hunk 11 above)
Comment (skinny85) [reply specifications/cloud_assembly.md]: +1 that "drop" for me here reads like a verb, not a noun. "Droplet" is a lot better, however it's already used by EC2... but maybe we don't care.

File: specifications/cloud_assembly.md [hunk 25]
```diff
@@ -0,0 +1,473 @@
...
+  + Plus: `+` (`0x2B`)
+  + Minus: `-` (`0x2D`)
+  + Forward-slash: `/` (`0x2F`)
+  + Underscore: `_` (`0x5F`)
+* A *Logical ID* **MUST NOT** contain the `.` (`0x2E`) character as it is used in the string substitution pattern for
+  cross-drop references to separate the *Logical ID* from the *attribute* name.
+
+In other words, *Logical IDs* are expected to match the following regular expression:
+```js
+/^[A-Za-z0-9+\/_-]{1,256}$/
```
Comment (skinny85) [inline specifications/cloud_assembly.md]: Can it be things like `-`, `+`, `_`, `0` then? Seems a little too permissive maybe?

File: specifications/cloud_assembly.md [hunk 26]
```diff
@@ -0,0 +1,473 @@
...
+
+### Drop
+Clouds are made of Drops. Thet are the building blocks of *Cloud Assemblies*. They model a part of the
+*cloud application* that can be deployed independently, provided its dependencies are fulfilled. Drops are specified
+using [JSON] objects that **MUST** conform to the following schema:
+
+Key          |Type                  |Required|Description
+-------------|----------------------|:------:|-----------
+`type`       |`string`              |Required|The [*Drop Type*](#drop-type) specifier of this Drop.
+`environment`|`string`              |required|The target [environment](#environment) in which Drop is deployed.
```
Comment (skinny85) [inline specifications/cloud_assembly.md]: "in which **this** Drop is deployed"?

File: specifications/cloud_assembly.md (hunk 25 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: Why? I have no problem with those. Their primary goal is not human-readability, it's supposed to be machine-processed. I find expressing "only letters and numbers, -, +, _, but must start by a letter" starts looking like the expression of a poor password policy... I don't want to impose restrictions that are convoluted to express in plain English unless I have a good reason to.

File: specifications/cloud_assembly.md (hunk 11 above)
Comment (RomainMuller) [reply specifications/cloud_assembly.md]: I don't think the collision with EC2 is so material (a search of the combined terms using a popular search engine was not super conclusive). There's other collisions that could be more problematic (but IANAL & I'm not going to discuss this here :D)

File: specifications/cloud_assembly.md (hunk 25 above)
Comment (skinny85) [reply specifications/cloud_assembly.md]: I don't think `0`, or `-`, or `+` are great logical IDs, but maybe it's just me. Feel free to ignore.

**Comment by rix0rrr** [comment]
What's going on with this PR?

**Comment by eladb** [comment]
Hold it off for now. We are meeting with CNAB next week and will be wiser after the meeting.

**Comment by RomainMuller** [comment]
Abandoning this for now (resolving without merge).

**Comment by hjander** [comment]
What will be the future direction regarding the Cloudlet/CNAB topic ?

**Comment by eladb** [comment]
@hjander thanks for querying - @RomainMuller left a [comment](https://github.com/awslabs/aws-cdk/issues/956#issuecomment-460583556) on #956 -- shall we continue the discussion there?
